zetaclient/hsm_signer.go

```diff
@@ -0,0 +1,172 @@
+package zetaclient
+
+import (
+	"os"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	clienttx "github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	"github.com/frumioj/crypto11"
+	"github.com/pkg/errors"
+	keystone "github.com/zeta-chain/keystone/keys"
+)
+
+const hsmPath = "HSM_PATH"
+const hsmPIN = "HSM_PIN"
+const hsmLabel = "HSM_LABEL"
+
+// HsmSign Generates signature of msg using the key indexed by the label through the HSM defined in the config
+func HsmSign(config *crypto11.Config, msg []byte, label string) (signature []byte, err error) {
+	keyring, err := keystone.NewPkcs11(config)
+	if err != nil {
+		return
+	}
+	key, err := keyring.Key(label)
+	if err != nil {
+		return
+	}
+	signature, err = key.Sign(msg, nil)
+	return
+}
+
+// GenerateKey This generates a new key using one of the supported algorithms and a label identifier through the HSM
+func GenerateKey(label string, algorithm keystone.KeygenAlgorithm, config *crypto11.Config) (*keystone.CryptoKey, error) {
+	keyring, err := keystone.NewPkcs11(config)
+	if err != nil {
+		return nil, err
+	}
+	key, err := keyring.NewKey(algorithm, label)
+	if err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+// GetHSMAddress This address is generated by secp256k1 curve from cosmos sdk
+func GetHSMAddress(config *crypto11.Config, label string) (types.Address, types.PubKey, error) {
+	keyring, err := keystone.NewPkcs11(config)
+	if err != nil {
+		return nil, nil, err
+	}
+	key, err := keyring.Key(label)
+	if err != nil {
+		return nil, nil, err
+	}
+	pubKey := key.PubKey()
+	return pubKey.Address(), pubKey, nil
+}
+
+// SignWithHSM signs a given tx with a named key.
+// // This is adapted from github.com/cosmos/cosmos-sdk/client/tx Sign() function; Modified to use an HSM.
+// // The resulting signature will be added to the transaction builder overwriting the previous
+// // ones if overwrite=true (otherwise, the signature will be appended).
+// // Signing a transaction with multiple signers in the DIRECT mode is not supported and will
+// // return an error.
+// // An error is returned upon failure.
```

Comment:
```suggestion
// This is adapted from github.com/cosmos/cosmos-sdk/client/tx Sign() function; Modified to use an HSM.
// The resulting signature will be added to the transaction builder overwriting the previous
// ones if overwrite=true (otherwise, the signature will be appended).
// Signing a transaction with multiple signers in the DIRECT mode is not supported and will
// return an error.
// An error is returned upon failure.
```